src/image/detail/MaskDict.cc

```diff
@@ -31,18 +32,48 @@ class MaskDict::GlobalState {
 
     using HandleList = std::set<std::weak_ptr<MaskDict>, std::owner_less<std::weak_ptr<MaskDict>>>;
 
-    static GlobalState & get() {
-        static GlobalState instance;
-        return instance;
-    }
+    /*
+     * RAII guard for GlobalState.
+     *
+     * Default-constructing a Lock instance is the only way to access
+     * GlobalState - this acquires GlobalState's mutex, ensuring only one
+     * thread has access to its attributes at one time.  The mutex is released
+     * when the Lock instance goes out of scope.
+     */
+    class Lock {
```

Comment:
That global acquisition and release of the lock isn't external; it's mandated by GlobalState's own API (in that Lock is the only external way to access that state).  I suppose someone really motivated could steal a raw pointer to its state and smuggle it out of the RAII block, but I'm not actually worried about that scenario.

Reply:
It's external in that a true thread-safe class manages any locks (or equivalent functionality) that it needs, rather than relying on a relationship with another class. I'm not worried about security so much as ease of comprehension, and as noted in the original comment doing things differently might be too inflexible.